File: dataset/human/c/file_9709.c
#include <stdio.h>
#define N 100000

typedef struct{
  int parent;
  int left;
  int right;
} Node;

int depth(int);
void judgenode(int);
void children(int);

Node node[N];

main() {
  int i,j,id,k,n,left,child;

  scanf("%d",&n);

  for(i=0; i<n; i++) {
    node[i].parent = -1;
    node[i].left = -1;
    node[i].right = -1;
  }

  for(i=0; i<n; i++) {
    scanf("%d%d",&id,&k);
    for(j=0; j<k; j++){
      scanf("%d",&child);
      if(j == 0)
	node[id].left = child;
      else
	node[left].right = child;
      node[child].parent = id;
      left = child;
    }
  }


  for(i=0; i<n; i++) {
    printf("node %d: parent = %d, depth = %d, ",i,node[i].parent,depth(i));
    judgenode(i);
    children(i);
  }

  return 0;

}


  int depth(int i) {
    int count=0, iparent;
    iparent=i;
    while(1) {
      if(node[iparent].parent == -1)
	break;
      iparent = node[iparent].parent;
      count+=1;
    }
    return count;
  }


  void judgenode(int i) {
    if(node[i].left == -1 && node[i].parent != -1)
      printf("leaf,");
    else if(node[i].parent == -1)
      printf("root,");
    else 
      printf("internal node,");
  }


void children(int i) {
  int sibling;
  printf(" [");
  if(node[i].left != -1) {
    printf("%d",node[i].left);
    sibling = node[i].left;
    while(1) {
      if(node[sibling].right == -1)
	break;
      printf(", %d",node[sibling].right);
      sibling = node[sibling].right;
    }
  }
  printf("]\n");
}

File: dataset/human/java/file_9709.java
/**
 * Tree - Binary Trees
 */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

import static java.lang.Integer.parseInt;

public class Main {

	static final int NIL = -1;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		//String line;

		int n = parseInt(br.readLine());
		Node[] T = new Node[n];

		StringTokenizer st;
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			int id = parseInt(st.nextToken());
			T[id] = new Node();
			T[id].node = id;
			T[id].left = parseInt(st.nextToken());
			T[id].right = parseInt(st.nextToken());
		}
		// set parent
		for (int i = 0; i < n; i++) {
			if (T[i].left != NIL) T[T[i].left].parent = i;
			if (T[i].right != NIL) T[T[i].right].parent = i;
		}
		// set depth
		int root = 0;
		while (T[root].parent != NIL) root++;
		Deque<Node> stack;

		stack = new ArrayDeque<Node>();
		stack.push(T[root]);
		while (!stack.isEmpty()) {
			Node u = stack.pop();
			if (u.left != NIL) {
				T[u.left].depth = u.depth + 1;
				stack.push(T[u.left]);
			}
			if (u.right != NIL) {
				T[u.right].depth = u.depth + 1;
				stack.push(T[u.right]);
			}
		}
		// set height
		stack = new ArrayDeque<Node>();
		stack.push(T[root]);
		while (!stack.isEmpty()) {
			Node u = stack.pop();
			if (u.height == -1) {
				u.height = 0;
				stack.push(u);
				if (u.left != NIL) stack.push(T[u.left]);
				if (u.right != NIL) stack.push(T[u.right]);
			} else {
				u.height = Math.max(
						u.left != NIL ? T[u.left].height + 1 : 0,
						u.right != NIL ? T[u.right].height + 1 : 0
				);
			}
		}
		// set sibling
		for (int i = 0; i < n; i++) {
			if (T[i].left != NIL && T[i].right != NIL) {
				T[T[i].left].sibling = T[i].right;
				T[T[i].right].sibling = T[i].left;
			}
		}
		// set degree
		for (int i = 0; i < n; i++) {
			if (T[i].left != NIL) T[i].degree++;
			if (T[i].right != NIL) T[i].degree++;
		}
		// set type
		for (int i = 0; i < n; i++) {
			if (T[i].parent == NIL) {
				T[i].type = "root";
			} else if (T[i].left != NIL || T[i].right != NIL) {
				T[i].type = "internal node";
			} else {
				T[i].type = "leaf";
			}
		}
		// print
		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < n; i++) {
			sb.append(T[i].toString());
		}
		System.out.print(sb.toString());
	} //end main
}

class Node {
	int node;
	int parent = -1;
	int left = -1;
	int right = -1;
	//
	int degree = 0;
	int sibling = -1;
	int depth = 0;
	int height = -1;
	String type;

	public String toString() {
		return String.format(
				"node %d: parent = %d, sibling = %d, degree = %d, depth = %d, height = %d, %s\n",
				node,
				parent,
				sibling,
				degree,
				depth,
				height,
				type);
	}
}
